File: BlueQuartz/5100WG/trunk/cce/client/php/class/CceClient.php
<?php
// Author: Harris Vargon-Lloyd, Kevin K.M. Chiu
// Copyright 2000, Cobalt Networks.  All rights reserved.
// $Id: CceClient.php 233 2003-08-07 07:39:51Z shibuya $

global $isCceClientDefined;
if($isCceClientDefined)
  return;
$isCceClientDefined = true;

include("CceError.php");
include("System.php");

class CceClient {
  //
  // private variables
  //

  var $handle;
  var $isConnected;

  //
  // public methods
  //

  // description: constructor
  function CceClient() {
    $this->handle = cce_new();

    // set default
    $this->isConnected = false;
  }

  // description: authenticate to the server
  // param: userName: user name in string
  // param: password: password in string
  // returns: session ID if succeed, empty string if failed
  function auth($userName, $password) {
    return cce_auth($this->handle, $userName, $password);
  }

  function authkey( $userName, $sessionId ) {
    return cce_authkey( $this->handle, $userName, $sessionId );
  }

  function whoami( ) {
    return cce_whoami( $this->handle );
  }

  function bye() {
    $this->isConnected = false;

    return cce_bye($this->handle);
  }

  function commit() {
    return cce_commit($this->handle);
  }

  function endkey() {
    return cce_endkey($this->handle);
  }

  // description: connect to CCE
  // param: socketPath: the path of the Unix domain socket to CCE
  // returns: true if succeed, false otherwise
  function connect($socketPath = "") {
    if($socketPath == "") {
      // get from config file
      $system = new System();
      $socketPath = $system->getConfig("ccedSocketPath");
    }

    $this->isConnected = cce_connect($this->handle, $socketPath);

    return $this->isConnected;
  }

  function create($class, $vars = array()) {
    return cce_create($this->handle, $class, $vars);
  }

  function destroy($oid) {
    return cce_destroy($this->handle, $oid);
  }

  function destroyObjects($class, $vars = array()) {
    $oids = $this->find($class, $vars);
    for($i = 0; $i < count($oids); $i++)
      $this->destroy($oids[$i]);
  }

  // description: get the last occured error
  // returns: an array of CceError objects
  function errors() {
    $errorObjs = array();

    $errors = cce_errors($this->handle);
    for($i = 0; $i < count($errors); $i++) {
      $error = $errors[$i];
      $errorObjs[] = new CceError($error["code"], $error["oid"], $error["key"], $error["message"], array("code" => $error["code"], "oid"  => $error["oid"], "key"  => $error["key"]));
    }

    return $errorObjs;
  }

  // description: returns array of hashes.  Each hash contains
  // information about a particular error, include "code", "oid",
  // "key", and "message".  
  function raw_errors() {
	return cce_errors($this->handle);
  }

  // the legacy find command
  function find($class, $vars = array()) {
    return cce_find($this->handle, $class, $vars, "", 0);
  }

  // find objects, returning a list sorted by the values in
  // the $key attribute
  function findSorted($class, $key, $vars = array()) {
	return cce_find($this->handle, $class, $vars, $key, 0);
  }

  // like findSorted, but sort is numeric instead of
  // alphabetic.
  function findNSorted($class, $key, $vars = array()) {
	return cce_find($this->handle, $class, $vars, $key, 1);
  }

  function get($oid, $namespace = "") {
    return cce_get($this->handle, $oid, $namespace);
  }

  // description: get a CCE object
  // returns: a property hash or null if the object cannot be found
  function getObject($class, $vars = array(), $namespace = "") {
    $oids = $this->find($class, $vars);
    if(count($oids) > 0)
      return $this->get($oids[0], $namespace);
    else
      return null;
  }

  // description: get an array of CCE objects
  // returns: an array of property hashes
  function getObjects($class, $vars = array(), $namespace = "") {
    $oids = $this->find($class, $vars);

    $objects = array();
    for($i = 0; $i < count($oids); $i++)
      $objects[] = $this->get($oids[$i], $namespace);

    return $objects;
  }

  // returns: true if the client is connected to the server, false otherwise
  function isConnected() {
    return $this->isConnected;
  }

  function names($arg) {
    return cce_names($this->handle, $arg);
  }

  function set($oid, $namespace = "", $vars = array()) {
    return cce_set($this->handle, $oid, $namespace, $vars);
  }

  // description: set a CCE object
  // returns: true on success, false otherwise
  function setObject($class, $setVars = array(), $namespace = "", $findVars = array()) {
    $oids = $this->find($class, $findVars);
    if(count($oids) > 0)
      return $this->set($oids[0], $namespace, $setVars);
    else
      return 0;
  }

  // description: set a CCE object. If it does not exist, create it
  function setObjectForce($class, $setVars = array(), $namespace = "", $findVars = array()) {
    $oids = $this->find($class, $findVars);

    // create if necessary
    if(count($oids) == 0)
      if(!$this->create($class))
	return 0;

    return $this->setObject($class, $setVars, $namespace, $findVars);
  }

  function array_to_scalar( $array ) {
		$result = "";
  	while( list($index,$val) = each($array) ) {
			# First quote quotes.
			$val = strtr($val,"\\","\\\\");
			# Then quote breaks,
			$val = strtr($val,"","\\:");
			$result .= $val . ':';
		}
		if( $val ) {
			chop($val);
		}
		return $val;
  }

	function scalar_to_array( $scalar ) {
		$ret_array = array();
		$array = explode(":",$scalar);

		while( list($index,$val) = each($array) ) {

			$elem .= $curr_elem;
	
			if( substr($elem, -1) != "\\" ) {
				$elem = strstr($elem,"\\\\","\\");
				array_push($ret_array, $elem);
			} else {
				chop($elem);
				$elem .= ":";
			}
		}

		if( $elem ) {
			array_push($ret_array, $elem);
		}

		return $ret_array;
	}

  // Description: find out whether or not this user has the capability $capName
  // returns: true if the user has this cce capability, false otherwise
  function getCapable( $capName) {
    if (!($cap = $this->getObject("Capability", $capName))) {
      print("Capability Object with namespace=$capName not found!");
      exit();
    }
    return $cap["capable"];
  }
}

/*
Copyright (c) 2003 Sun Microsystems, Inc. All  Rights Reserved.

Redistribution and use in source and binary forms, with or without modification, 
are permitted provided that the following conditions are met:

-Redistribution of source code must retain the above copyright notice, this  list of conditions and the following disclaimer.

-Redistribution in binary form must reproduce the above copyright notice, 
this list of conditions and the following disclaimer in the documentation and/or 
other materials provided with the distribution.

Neither the name of Sun Microsystems, Inc. or the names of contributors may 
be used to endorse or promote products derived from this software without 
specific prior written permission.

This software is provided "AS IS," without a warranty of any kind. ALL EXPRESS OR IMPLIED CONDITIONS, REPRESENTATIONS AND WARRANTIES, INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT, ARE HEREBY EXCLUDED. SUN MICROSYSTEMS, INC. ("SUN") AND ITS LICENSORS SHALL NOT BE LIABLE FOR ANY DAMAGES SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING OR DISTRIBUTING THIS SOFTWARE OR ITS DERIVATIVES. IN NO EVENT WILL SUN OR ITS LICENSORS BE LIABLE FOR ANY LOST REVENUE, PROFIT OR DATA, OR FOR DIRECT, INDIRECT, SPECIAL, CONSEQUENTIAL, INCIDENTAL OR PUNITIVE DAMAGES, HOWEVER CAUSED AND REGARDLESS OF THE THEORY OF LIABILITY, ARISING OUT OF THE USE OF OR INABILITY TO USE THIS SOFTWARE, EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.

You acknowledge that  this software is not designed or intended for use in the design, construction, operation or maintenance of any nuclear facility.
*/
?>

File: BlueQuartz/5100R/tags/raq550_OSS_1_0/ui/base-scandetection.mod/src/commonlib/CceClient.c
/*
** Copyright 2001 Sun Microsystems, Inc. All rights reserved.
** $Id: CceClient.c,v 1.5.2.2 2002/02/20 18:01:10 ge Exp $
*/

/*
 * This code is a hodgepodge to get email alerts to be sent out i18n
 * style.  The first part comes from the swatch+cce code and the second
 * part puts it all together.  I choose to use this client because its
 * i18n compliant and ours wasn't.  
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <ctype.h>

//#include <glib.h>
#include <cce/cce.h>
#include <cce/i18n.h>

#include "CceClient.h"
#include "libphoenix.h"

i18n_handle *i18n = NULL;
cce_handle_t *cce = NULL;

/***********************************************************************/

#define DEF_LOCALE      "en"

/* where to look for the sys locale */
#define SYS_LOCALE_FILE "/etc/cobalt/locale"

#define DOMAIN    "base-scandetection"

/* how to send mail */
#define MAIL_CMD  "/usr/sausalito/bin/i18nmail.pl"
#define MAIL_SUBJ "alertEmailSubject"
#define MAIL_BODY "alertEmailBody"


/*
 *  This code was mercilessly stolen from swatch+cce
 */
static int
send_mail(char *emails, char **messages, char* locale, i18n_vars *v) 
{
	char *cmd;
	char *tmpfile;
	FILE *fp;
	int fd;
	int len;
	char *subj;

#if 0	
	if (!messages[0] || !*emails) {
#endif
	if (!*emails) {
		/* no messages to report, or no one to report them to */
		return 0;
	}

	/* get and open a temp file */
	tmpfile = strdup("/tmp/.alertd-XXXXXX");
	fd = mkstemp(tmpfile);
	if (fd < 0) {
	  LOG(LOG_ERROR, "send_mail: mkstemp(): %s\n", strerror(errno));
	  return -1;
	}
	fp = fdopen(fd, "w");
	if (!fp) {
		LOG(LOG_ERROR, "send_mail: fdopen(): %s\n", strerror(errno));
		unlink(tmpfile);
		free(tmpfile);
		return -1;
	}

	/* put the i18ned body */
#if 0
	v = i18n_vars_new();
#endif
	fprintf(fp, "%s\n\n", i18n_get(i18n, MAIL_BODY, DOMAIN, v));
#if 0
	i18n_vars_destroy(v);
	/* write each message */
	for (i = 0; messages[i]; i++) {
		fprintf(fp, "* %s\n", messages[i]);
	}
#endif
	fclose(fp);

	/* i18n the subject */
#if 0
	v = i18n_vars_new();
#endif
	subj = i18n_get(i18n, MAIL_SUBJ, DOMAIN, v);
#if 0
	18n_vars_destroy(v);
#endif
	/* shut Perl up */
	setenv("PERL_BADLANG", "0", 1);

	/* "MAIL_CMD -s "subj" emails < tmpfile" */
	/* OR, if locale is specified, 
	   "MAIL_CMD -l locale -s "subj" emails < tmpfile" */
	len = strlen(MAIL_CMD) + 3 + 2 + strlen(subj) + 2 
		+ strlen(emails) + 3 + strlen(tmpfile) + 1;
	if (locale) {
	  len += 4 + strlen(locale);
	}

	cmd = malloc(len);
	if (!cmd) {
		LOG(LOG_ERROR, "send_mail: malloc(): %s\n", strerror(errno));
		unlink(tmpfile);
		free(tmpfile);
		return -1;
	}

	if (locale) {
	  snprintf(cmd, len, "%s -l %s -s \"%s\" %s < %s", MAIL_CMD, locale,
		   subj, emails, tmpfile);
	} else {
	  snprintf(cmd, len, "%s -s \"%s\" %s < %s", MAIL_CMD, 
		   subj, emails, tmpfile);
	}
	LOG(LOG_DEBUG, "sending mail (%s)\n", cmd);

	if (system(cmd)) {
		LOG(LOG_ERROR, "unexpected mailer error (%s)\n", strerror(errno));
		unlink(tmpfile);
		free(tmpfile);
		return -1;
	}

	unlink(tmpfile);
	free(tmpfile);

	return 0;
}



/* this is another of those static-memory-returning functions.  
 * freeing it's return would be a bad idea
 */
static char *
get_sys_locale(void)
{
	int fd,r;
	static char buf[6];

	fd=open(SYS_LOCALE_FILE, O_RDONLY);
	if (fd < 0) {
		snprintf(buf, sizeof(buf), "%s", DEF_LOCALE);
	}

	r=read(fd, buf, sizeof(buf)-1);
	buf[r]='\0';

	return buf;
}


/* FIXME: do better cleanup in here */
static char *
get_adm_locales(cce_handle_t *cce)
{
	cce_props_t *props;
	cscp_oid_t adm_oid;
	GSList *oidlist;

	props = cce_props_new();
	if (!props) {
		return NULL;
	}

	cce_props_set(props, "name", "admin");
	oidlist = cce_find_cmnd(cce, "User", props);
	cce_props_destroy(props);
	if (!oidlist) {
		return NULL;
	}
	
	adm_oid = (cscp_oid_t)oidlist->data;

	props = cce_get_cmnd(cce, adm_oid, "");
	if (!props) {
		return NULL;
	}

	return cce_props_get(props, "localePreference");
}

/************************************************************************/

/* 
 * Sample packets - break them down into their components
 *
 * eth0:restrict: tcp 172.25.91.7/22 -> 172.25.91.1/2238 280 (3)
 * eth0:portscan: 3/1/icmp 172.25.91.7 <- 129.148.1.37 56 (2)
 */

#define NEXT_TOKEN {  while ( isspace(*p) )   \
                        p++;                  \
                      q = p;                  \
                      while (! isspace(*p) )  \
                        p++;                  \
                      *p++ = '\0';            \
                   };

static void
parse_alert(const char *message, i18n_vars *v)
{
    char msg[256];
    char *p, *q, *r;
    char *src, *dst;

    p = strncpy(msg, message, sizeof(msg));

    NEXT_TOKEN;
    /* LOG(LOG_DEBUG, "Iface=%s\n", q); */
    if ((r = strchr(q, ':')) != NULL)
        *r = '\0';
    i18n_vars_add(v, "interface", q);

    NEXT_TOKEN;
    /* LOG(LOG_DEBUG, "Protocol=%s\n", q); */
    i18n_vars_add(v, "protocol", q);
    
    NEXT_TOKEN;
    /* LOG(LOG_DEBUG, "IP1=%s\n", q); */
    src = q;

    NEXT_TOKEN;
    /* LOG(LOG_DEBUG, "Direction=%s\n", q); */
    r = q;

    NEXT_TOKEN;
    /* LOG(LOG_DEBUG, "IP2%s\n", q); */
    dst = q;

    /* Based on direction, setup fields */
    if (! strcmp(r, "->")) {
        i18n_vars_add(v, "direction", i18n_get(i18n, "outbound", DOMAIN, v));
    } else {
        char *t = src;
	src = dst;
	dst = t;
        i18n_vars_add(v, "direction", i18n_get(i18n, "inbound", DOMAIN, v));
    }

    /* Split the ip/port.  Icmps don't have port so leave empty. */
    if ((r = strchr(src, '/')) != NULL) {
        *r = '\0';
	i18n_vars_add(v, "srcport", r+1);
    } else {
        i18n_vars_add(v, "srcport", "");
    }
    i18n_vars_add(v, "srcaddr", src);

    if ((r = strchr(dst, '/')) != NULL) {
        *r = '\0';
	i18n_vars_add(v, "dstport", r+1);
    } else {
        i18n_vars_add(v, "dstport", "");
    }
    i18n_vars_add(v, "dstaddr", dst);

    NEXT_TOKEN;
    /* LOG(LOG_DEBUG, "Size=%s\n", q); */
    i18n_vars_add(v, "pktsize", q);
}

/* cce_connect() is defined in cce.h */

static int
cce_connect_local(void)
{
    if (! (cce = cce_handle_new()) ) {
        LOG(LOG_BASE, "cannot get CCE handle");
	return 0;
    }

    if (!cce_connect_cmnd(cce, NULL)) {
        LOG(LOG_BASE, "cannot connect to CCE");
	cce_handle_destroy(cce);
	return 0;
    }
    return 1;
}

static void
cce_bye(void)
{
    cce_bye_cmnd(cce);
    cce_handle_destroy(cce);
}

static int
i18n_connect(void)
{
    char *locale;

    /* set up i18n */
    locale=get_adm_locales(cce);
    if(strcmp(locale,"browser")==0){
        locale=get_sys_locale();
    }
    i18n = i18n_new(DOMAIN, locale);
    if (!i18n) {
	LOG(LOG_BASE, "cannot initialize i18n");
	return 0;
    }
    return 1;
}

/* 
 * This code is pretty messy.  I guess that happens when two disjoint
 * pieces of code try to merge together.
 */
int send_alert(
	       time_t timestamp,
	       const char *to,
	       const char *subject,
	       const char *message
	       )
{
    char *subj_i18n;
    i18n_vars *v;
    LOG(LOG_BASE, "Sending alert to %s", to);

    if (! cce_connect_local())
        return -1;

    if (! i18n_connect())
        return -1;

    /* The subject is an i18n tag; we look it up, take its value and add
       to the vars array to get substituted in the email. */
    v = i18n_vars_new();
    subj_i18n = i18n_get(i18n, (char *)subject, DOMAIN, v);
    /* This replaces [[VAR.variable]] in .po file */
    i18n_vars_add(v, "subject", subj_i18n);
    i18n_vars_add(v, "timestamp", i18n_get_datetime(i18n, timestamp));
    i18n_vars_add(v, "type", subj_i18n);
    i18n_vars_add(v, "logentry", (char *) message);
    parse_alert(message, v);
    send_mail((char *)to, NULL, NULL, v);

#if 0
    // Example property change
    props = cce_props_new();
    cce_props_set(props, "monitor", "987");
    cce_set_cmnd(cce, oid, "Fwall", props);
    cce_props_destroy(props);
#endif

    i18n_vars_destroy(v);
    i18n_destroy(i18n);
    cce_bye();

    return 0;
}

#ifdef UNDEF

/* NO MORE CCE ACCESS FROM C PROGRAMS! */

/*
 *  get_admin_email - look in CCE and return admin email list, space separated
 *
 *  Args:    buffer and size
 *  Returns: true if successful, false otherwise
 */               
int get_admin_email(char *buffer, int size)
{
    cce_props_t *props;
    cscp_oid_t oid;
    GSList *oids, *p;
    char *emails, *q;
    int i, remainder = size;

    if (! cce_connect_local())
        return(0);

    if ((oids = cce_find_cmnd(cce, "System", NULL)) == NULL) {
        LOG(LOG_BASE, "Unable to get System from CCE");
	cce_bye();
	return(0);
    }

    oid = (cscp_oid_t)oids->data;   
    props = cce_get_cmnd(cce, oid, "Scandetection");

    emails = cce_props_get(props, "alertEmail");
   
    snprintf(buffer, strlen(emails) +1 , "%s", emails); 
    cce_bye();
    return( strlen(buffer) );
}

/*
 *  isPortScanOn
 * 
 * returns -1 when System.Scandetection isn't found - fwall not installed
 * returns 0 if System.Scandetection.paranoiaLevel == (0 || 1)
 * returns 1 if System.Scandetection.paranoiaLevel == 2
 */               
int isPortScanOn(void)
{
    cce_props_t *props;
    cscp_oid_t oid;
    GSList *oids, *p;
    char *paranoiaLevelChar;
    char *alertMeChar;
    int paranoiaLevel, alertMe;

    if (! cce_connect_local())
        return(-1);

    if ((oids = cce_find_cmnd(cce, "System", NULL)) == NULL) {
        LOG(LOG_BASE, "Unable to get System from CCE");
	cce_bye();
	return(-1);
    }

    oid = (cscp_oid_t)oids->data;   
    props = cce_get_cmnd(cce, oid, "Scandetection");

    paranoiaLevelChar = cce_props_get(props, "paranoiaLevel");	 
    alertMeChar = cce_props_get(props, "alertMe");
//LOG(LOG_BASE, "PAR=%s ALERT=%s", paranoiaLevelChar, alertMeChar);

    paranoiaLevel = (int)paranoiaLevelChar[0] - 48;
    alertMe = (int)alertMeChar[0] - 48;

    if (paranoiaLevel > 0 && alertMe > 0) {
	cce_bye();
//LOG(LOG_BASE, "RETURNING TRUE");
	return(1);
    }    


    cce_bye();
//LOG(LOG_BASE, "RETURNING FALSE");
    return(0);
}
#endif /*UNDEF*/
/* Copyright (c) 2003 Sun Microsystems, Inc. All  Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met: 
 * 
 * -Redistribution of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * 
 * -Redistribution in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution. 
 *
 * Neither the name of Sun Microsystems, Inc. or the names of contributors may
 * be used to endorse or promote products derived from this software without 
 * specific prior written permission.

 * This software is provided "AS IS," without a warranty of any kind. ALL EXPRESS OR IMPLIED CONDITIONS, REPRESENTATIONS AND WARRANTIES, INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT, ARE HEREBY EXCLUDED. SUN MICROSYSTEMS, INC. ("SUN") AND ITS LICENSORS SHALL NOT BE LIABLE FOR ANY DAMAGES SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING OR DISTRIBUTING THIS SOFTWARE OR ITS DERIVATIVES. IN NO EVENT WILL SUN OR ITS LICENSORS BE LIABLE FOR ANY LOST REVENUE, PROFIT OR DATA, OR FOR DIRECT, INDIRECT, SPECIAL, CONSEQUENTIAL, INCIDENTAL OR PUNITIVE DAMAGES, HOWEVER CAUSED AND REGARDLESS OF THE THEORY OF LIABILITY, ARISING OUT OF THE USE OF OR INABILITY TO USE THIS SOFTWARE, EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 * 
 * You acknowledge that  this software is not designed or intended for use in the design, construction, operation or maintenance of any nuclear facility.
 */
